apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -227,7 +227,16 @@ public void onChange(ConfigurationOption<?> configurationOption, Double oldValue
     }
 
     public Transaction noopTransaction() {
-        return transactionPool.createInstance().startNoop();
+        return createTransaction().startNoop();
+    }
+
+    private Transaction createTransaction() {
+        Transaction transaction = transactionPool.createInstance();
+        while (transaction.getReferenceCount() != 0) {
+            logger.warn("Tried to start a transaction with a non-zero reference count {} {}", transaction.getReferenceCount(), transaction);
```

Comment:
If the reference count is higher than 0 the transaction would not have been recycled. That can happen if we miss to deactivate, for example. But in this case, we have the safety net of the GC which will collect the instance if it's not reachable anymore.

But it could also happen that we deactivate again when the reference count is already 0 and we have already recycled the transaction. The pattern would be

```java
t.activate();
t.deactivate();
t.deactivate();
```

That actually happened before https://github.com/elastic/apm-agent-java/pull/605. In this case the reference count is also not 0. It's better not to try reusing this instance again as it might still be in use. So we would just let GC take care of this transaction which collects it once not referenced anymore.

Bottom line - I don't think there is an increased risk of memory leaks.